### ISSUE TITLE
Kotest is very slow to start.

### ISSUE THREAD

**Issue by tonicsoft** [kotest/kotest#3126]
This is a follow up of https://github.com/kotest/kotest/issues/2765. The problem discovered via that issue was a classpath scanning issue, for which a workaround is now provided via system properties:

```
kotest.framework.classpath.scanning.config.disable=true
kotest.framework.config.fqn=com.wibble.KotestConfig
```

Unfortunately the workaround doesn't work when using the IDE, and I have raised that issue separately here: https://github.com/kotest/kotest/issues/3125

However when using the command line, the workaround does speed up Kotest. However, Kotest remains several times slower than Junit. Please see https://github.com/tonicsoft/slow-kotest for an example project that can be used to directly compare junit with kotest.

From my measurements, junit could run the test in <1 second, and kotest more like 3 seconds.

**Comment by sksamuel** [comment]
So what you need is a couple of things. Firstly, disable classpath scanning of `autoscan` in addition to config. Then your timing script is the same for me in junit vs kotest.

```
~/development/workspace/kotest/slow-kotest (master ✗) ./collect-timings.sh

> Task :junit:test

FastJunitTest > fastTest() PASSED

BUILD SUCCESSFUL in 856ms
2 actionable tasks: 1 executed, 1 up-to-date
0.75user 0.05system 0:01.25elapsed 64%CPU (0avgtext+0avgdata 92008maxresident)k
0inputs+64outputs (8major+18031minor)pagefaults 0swaps

> Task :kotest:test

SlowKoTest > kotest > SlowKoTest.should take about 4 seconds to run this test on a top end laptop PASSED

BUILD SUCCESSFUL in 1s
2 actionable tasks: 1 executed, 1 up-to-date
0.87user 0.06system 0:01.60elapsed 58%CPU (0avgtext+0avgdata 97904maxresident)k
0inputs+64outputs (8major+20470minor)pagefaults 0swaps
```

Secondly, to make this work via the intellij plugin, you need to specify the props in a `kotest.properties` file. I have updated the docs to clarify this [here](https://kotest.io/docs/next/framework/project-config.html) and [here](https://kotest.io/docs/next/intellij/intellij-properties.html).

Let me know if you think it's still not clear enough.

**Comment by tonicsoft** [comment]
That's awesome, very helpful additions to the docs too.

Thanks, and apologies for the rtfm.

**Comment by Fryie** [comment]
Disabling classpath scanning reduced one of our builds by 20 minutes.

Maybe this should be a default setting?

**Comment by sksamuel** [comment]
It will be the default in 6.0

**Comment by tonisives** [comment]
I want to again report on this/similar issue. For me, there is about a ~10s lag between the `0 tests completed` and the actual start of the tests.

![out](https://github.com/kotest/kotest/assets/1083534/ecd472e2-af37-4c1e-aeeb-a4d204a7b684)


I have parallelism = 5. There are only 31 tests. These are all fast tests.

I have the scanning disabled in kotest.properties

```
kotest.framework.classpath.scanning.config.disable=true
kotest.framework.classpath.scanning.autoscan.disable=true
``` 

Is this 10s lag normal behaviour? What could be the reason for it?

`kotest 5.7.2`, `kotlin 1.9.0`

**Comment by sksamuel** [comment]
Is this a public project I can clone? Or do you have a reproducer ?

**Comment by tonisives** [comment]
> Is this a public project I can clone? Or do you have a reproducer ?

I created a repo with my test results:https://github.com/tonisives/kotest-performance

It seems 1-2s faster now that I removed a bunch of client's code.

**Comment by NersesAM** [comment]
I don't know what changed in 5.7.0 but upgrading from 5.6.2 we've gone from 2m build time to 20m.
I tried looking at changes but couldn't figure out what change actually could have caused it. 

surefire forked jvm main thread is waiting on 

```
Stack Trace	Count	Percentage
void jdk.internal.misc.Unsafe.park(boolean, long)	17	100 %
void java.util.concurrent.locks.LockSupport.park(Object)	17	100 %
int java.util.concurrent.FutureTask.awaitDone(boolean, long)	17	100 %
Object java.util.concurrent.FutureTask.get()	17	100 %
ScanResult io.github.classgraph.ClassGraph.scan(ExecutorService, int)	17	100 %
ScanResult io.github.classgraph.ClassGraph.scan(int)	17	100 %
ScanResult io.github.classgraph.ClassGraph.scan()	17	100 %
void io.kotest.engine.config.ApplyConfigFromAutoScanKt.applyConfigFromAutoScan(ProjectConfiguration)	9	52.9 %
Object io.kotest.engine.config.ConfigManager.compile-gIAlu-s(ProjectConfiguration, Function0)	9	52.9 %
ProjectConfiguration io.kotest.engine.config.ConfigManager.initialize(ProjectConfiguration, Function0)	9	52.9 %
KotestEngineDescriptor io.kotest.runner.junit.platform.KotestJunitPlatformTestEngine.discover(EngineDiscoveryRequest, UniqueId)	9	52.9 %
TestDescriptor io.kotest.runner.junit.platform.KotestJunitPlatformTestEngine.discover(EngineDiscoveryRequest, UniqueId)	9	52.9 %
TestDescriptor org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(TestEngine, LauncherDiscoveryRequest, LauncherDiscoveryListener, Function)	9	52.9 %
Map org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(LauncherDiscoveryRequest, EngineDiscoveryOrchestrator$Phase, LauncherDiscoveryListener, Function)	9	52.9 %
Map org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(LauncherDiscoveryRequest, EngineDiscoveryOrchestrator$Phase, Function)	9	52.9 %
LauncherDiscoveryResult org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(LauncherDiscoveryRequest, EngineDiscoveryOrchestrator$Phase)	9	52.9 %
LauncherDiscoveryResult org.junit.platform.launcher.core.DefaultLauncher.discover(LauncherDiscoveryRequest, EngineDiscoveryOrchestrator$Phase)	9	52.9 %
TestPlan org.junit.platform.launcher.core.DefaultLauncher.discover(LauncherDiscoveryRequest)	9	52.9 %
TestPlan org.junit.platform.launcher.core.DelegatingLauncher.discover(LauncherDiscoveryRequest)	9	52.9 %
TestPlan org.apache.maven.surefire.junitplatform.LazyLauncher.discover(LauncherDiscoveryRequest)	9	52.9 %
boolean org.apache.maven.surefire.junitplatform.TestPlanScannerFilter.accept(Class)	9	52.9 %
TestsToRun org.apache.maven.surefire.api.util.DefaultScanResult.applyFilter(ScannerFilter, ClassLoader)	9	52.9 %
TestsToRun org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.scanClasspath()	9	52.9 %
RunResult org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(Object)	9	52.9 %
void org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess()	9	52.9 %

```

while there are massive number of these threads created

<img width="201" alt="image" src="https://github.com/kotest/kotest/assets/1935500/48c5d4c9-8ccc-49bf-baea-dc590ce8f8dd">

and GC is going crazy as these threads create GBs of garbage.

setting these in kotest.properties fixed it.
```
kotest.framework.classpath.scanning.config.disable=true
kotest.framework.classpath.scanning.autoscan.disable=true
```

**Comment by sksamuel** [comment]
Can you try 5.7.2

**Comment by NersesAM** [comment]
> Can you try 5.7.2

it is the same. The issue starts from 5.7.0

**Comment by sksamuel** [comment]
Duplicate of https://github.com/kotest/kotest/issues/3746

**Comment by GregYeo** [comment]
SpringBoot user would be failing to execute test when the test class constructor requires @Inject or @Autowire.
It can be fixed by passing constructor extension and setting config fqn directly without scanning.
As a result of static configuration start up, it reduces dramatically kotest starting time.

```properties
#src/test/resources/kotest.properties
kotest.framework.classpath.scanning.config.disable=true
kotest.framework.classpath.scanning.autoscan.disable=true
kotest.framework.config.fqn=com.example.KotestProjectConfig
```

```kotlin
package com.example

import io.kotest.core.config.AbstractProjectConfig
import io.kotest.core.extensions.Extension
import io.kotest.extensions.spring.SpringAutowireConstructorExtension

object KotestProjectConfig : AbstractProjectConfig() {
    override fun extensions(): List<Extension> {
        return listOf(
            SpringAutowireConstructorExtension
        )
    }
}
```

**Comment by sksamuel** [comment]
That is exactly the way to do it. Also this is made much clearer in the upcoming 6.0 release where classpath scanning is gone.
